File: frontend/src/sim/genetics/genotype.ts
import { controller, type Genome, type BrainState } from "../controller";
import { type Config } from "../config";
import { type RandomState } from "../random";
import { combineVector, meanVector, mutateVector } from "./operators";
import { BODY_PARTS } from "../body";

export interface Chromosome {
  readonly behavior: Genome;
  readonly physical: Float32Array;
}
export interface Genotype {
  readonly chromosomes: readonly Chromosome[];
}
const expressed = new WeakMap<Genotype, Chromosome>();
export function express(genotype: Genotype): Chromosome {
  const cached = expressed.get(genotype);
  if (cached) return cached;
  const [a, b] = genotype.chromosomes;
  const value = b
    ? {
        behavior: controller.express(a.behavior, b.behavior),
        physical: meanVector(a.physical, b.physical),
      }
    : a;
  expressed.set(genotype, value);
  return value;
}
export function seedGenotype(config: Config): Genotype {
  return {
    chromosomes: Array.from({ length: config.ploidy === "haploid" ? 1 : 2 }, () => ({
      behavior: controller.seed(),
      physical: new Float32Array(BODY_PARTS.length),
    })),
  };
}
function gamete(genotype: Genotype, rng: RandomState, config: Config): Chromosome {
  const [a, b] = genotype.chromosomes;
  return {
    behavior: controller.recombine(a.behavior, b.behavior, rng, config.crossover),
    physical: combineVector(a.physical, b.physical, rng, config.crossover),
  };
}
export const transmissions = {
  clonal: (parent: Genotype) => parent.chromosomes,
  selfing: (parent: Genotype, rng: RandomState, config: Config) => [
    gamete(parent, rng, config),
    gamete(parent, rng, config),
  ],
};
export function inherit(parent: Genotype, rng: RandomState, config: Config, state?: BrainState) {
  const acquired =
    state && config.learning === "plastic"
      ? assimilate(parent, state, config.learningRetention)
      : parent;
  const learned = controller.genomeDistance(express(parent).behavior, express(acquired).behavior);
  const base = { chromosomes: transmissions[config.transmission](acquired, rng, config) };
  const genome = mutateGenotype(base, rng, config);
  return {
    genome,
    mutated: !sameGenotype(base, genome),
    recombined: config.transmission === "selfing",
    learned,
  };
}
function assimilate(parent: Genotype, state: BrainState, retention: number): Genotype {
  const expressed = express(parent).behavior;
  return {
    chromosomes: parent.chromosomes.map((c) => ({
      physical: c.physical.slice(),
      behavior: controller.assimilate(c.behavior, expressed, state, retention),
    })),
  };
}
function mutateGenotype(genotype: Genotype, rng: RandomState, config: Config): Genotype {
  const chromosomes = genotype.chromosomes;
  return {
    chromosomes: chromosomes.map((c) => ({
      behavior: controller.mutate(
        c.behavior,
        rng,
        config.mutationRate,
        config.mutationScale,
        config.mutationKind
      ),
      physical: mutateVector(
        c.physical,
        rng,
        {
          rate: config.physicalMutationRate,
          scale: config.physicalMutationScale,
          kind: config.mutationKind,
        },
        3
      ),
    })),
  };
}
export function sameGenotype(a: Genotype, b: Genotype): boolean {
  return (
    a.chromosomes.length === b.chromosomes.length &&
    a.chromosomes.every(
      (c, i) =>
        controller.genomeDistance(c.behavior, b.chromosomes[i].behavior) === 0 &&
        c.physical.every((v, j) => v === b.chromosomes[i].physical[j])
    )
  );
}

File: engine/src/genetics/genotype.rs
use super::*;

impl Genotype {
    pub fn seed(c: &Config, chemistry: &Chemistry) -> Self {
        let allele = Chromosome {
            behavior: controller::seed(),
            physical: [0.; crate::organism::STOCKS],
            chemistry: Machinery::seed(chemistry, &c.source_species),
        };
        let mut g = Self {
            id: 1,
            parent: None,
            born: 0,
            learned: 0.,
            mutated: false,
            chromosomes: vec![allele; if c.ploidy == "diploid" { 2 } else { 1 }],
            compiled: None,
        };
        g.compile(c, chemistry);
        g
    }
    pub fn express(&self) -> Chromosome {
        let a = &self.chromosomes[0];
        let Some(b) = self.chromosomes.get(1) else {
            return a.clone();
        };
        Chromosome {
            behavior: controller::express(&a.behavior, &b.behavior),
            physical: std::array::from_fn(|i| (a.physical[i] + b.physical[i]) * 0.5),
            chemistry: Machinery::express(&a.chemistry, &b.chemistry),
        }
    }
    pub fn compile(&mut self, c: &Config, chemistry: &Chemistry) {
        let chromosome = self.express();
        let m = &chromosome.chemistry;
        let ratios = [
            1.,
            c.motor_ratio,
            c.storage_ratio,
            c.receptor_ratio,
            c.receptor_ratio,
            c.receptor_ratio,
            c.receptor_ratio,
            c.transporter_ratio,
            c.transporter_ratio,
            c.transporter_ratio,
            c.transporter_ratio,
            c.enzyme_ratio,
            c.enzyme_ratio,
            c.enzyme_ratio,
            c.enzyme_ratio,
            c.receptor_ratio,
            c.enzyme_ratio,
            c.enzyme_ratio,
            c.enzyme_ratio,
            c.enzyme_ratio,
        ];
        let ratios: Vec<_> = ratios.into_iter().chain([c.transporter_ratio; 2]).collect();
        let core = c.birth_mass * (chromosome.physical[0] as f64).exp();
        let mut body = std::array::from_fn(|i| {
            if i == 0 {
                core
            } else {
                core * ratios[i] * (1. + chromosome.physical[i] as f64).max(0.)
            }
        });
        for (slot, active) in m.programs.iter().enumerate() {
            if !active {
                body[crate::organism::enzyme_stock(slot)] = 0.;
            }
        }
        let operators = if let Some(previous) = &self.compiled {
            let mut op = previous.operators.clone();
            op.update(&previous.chromosome.chemistry, m, c, chemistry);
            op
        } else {
            crate::chemical_operators::Operators::compile(m, c, chemistry)
        };
        self.compiled = Some(Compiled {
            sequence_fingerprint: {
                use std::hash::{Hash, Hasher};
                let mut hash = std::collections::hash_map::DefaultHasher::new();
                postcard::to_stdvec(&self.chromosomes)
                    .unwrap()
                    .hash(&mut hash);
                hash.finish()
            },
            chromosome,
            body,
            operators,
            observation: Default::default(),
        });
    }
    pub fn inherit(
        &self,
        id: u64,
        tick: u64,
        state: &controller::State,
        rng: &mut Random,
        c: &Config,
        chemistry: &Chemistry,
    ) -> Self {
        let expressed = &self.compiled.as_ref().unwrap().chromosome.behavior;
        let mut acquired = self.chromosomes.clone();
        if c.learning == "plastic" {
            for a in &mut acquired {
                a.behavior =
                    controller::assimilate(&a.behavior, expressed, state, c.learning_retention);
            }
        }
        let learned =
            controller::genome_distance(&self.chromosomes[0].behavior, &acquired[0].behavior);
        if c.transmission == "selfing" {
            acquired = (0..2)
                .map(|_| Chromosome {
                    behavior: controller::recombine(
                        &acquired[0].behavior,
                        &acquired[1].behavior,
                        rng,
                        &c.crossover,
                    ),
                    physical: controller::combine(
                        &acquired[0].physical,
                        &acquired[1].physical,
                        rng,
                        &c.crossover,
                    )
                    .try_into()
                    .unwrap(),
                    chemistry: Machinery::combine(
                        &acquired[0].chemistry,
                        &acquired[1].chemistry,
                        rng,
                        &c.crossover,
                    ),
                })
                .collect();
        }
        let mut mutated = false;
        for a in &mut acquired {
            mutated |= controller::mutate(&mut a.behavior, rng, c);
            mutated |= controller::mutate_vector(
                &mut a.physical,
                rng,
                c.physical_mutation_rate,
                c.physical_mutation_scale,
                3.,
            );
            mutated |= a.chemistry.mutate(rng, c);
        }
        let mut child = Self {
            id,
            parent: Some(self.id),
            born: tick,
            learned,
            mutated,
            chromosomes: acquired,
            compiled: self.compiled.clone(),
        };
        child.compile(c, chemistry);
        child
    }
    pub fn validate(&self, c: &Config) -> Result<(), String> {
        if self.chromosomes.len() != if c.ploidy == "diploid" { 2 } else { 1 } {
            return Err("Invalid chromosome count".into());
        }
        for a in &self.chromosomes {
            controller::validate(&a.behavior)?;
            if a.physical.iter().any(|x| !x.is_finite() || x.abs() > 3.) {
                return Err("Invalid physical loci".into());
            }
            a.chemistry.validate()?;
            if !a.chemistry.programs.iter().any(|v| *v) {
                return Err("An enzyme template must remain".into());
            }
        }
        Ok(())
    }
}
